### ISSUE TITLE
Issue with 'Like' Matcher

### ISSUE THREAD

**Issue by Mukheem** [reecetech/pactman#78]
**Python 3.8.0 with Pactman 2.25.0**

When using Like(1) in response, We see it is returning an Object making the assertion fail.

As per Documentation, It is not checking the Data type.

Sample Code:

    def test_get_MS(self):
        # Provider Start
       expected = { 'skip': 0, 'limit': Like(150)}
       pact.given(
            'User ML exists and is not an administrator'
        ).upon_receiving(
            'a request for ML'
        ).with_request(
            method='GET',
            path='/api/*************',
            body={},
            headers={'Content-Type': 'application/json'},
        ).will_respond_with(200, body=expected )
        # Provider End
        
      # Consumer Start
        with pact:
            result = requests.get('http://**************')

        self.assertEqual(result.json(), expected)
        # Consumer End.

Result is attached in here.
self = <test_pactmane2e.GetUserInfoContract testMethod=test_get_MS>
first = {'from': '2020-01-21T12:30:00+11:00', 'limit': 150, 'presenceChanges': [], 'skip': 0, ...}
second = {'from': '2020-01-21T12:30:00+11:00', 'limit': **<pactman.mock.matchers.Like object at 0x0000023A79CAA400>**, 'presenceChanges': [], 'skip': 0, ...}
msg = None

![image](https://user-images.githubusercontent.com/19348206/73722401-683aec00-474c-11ea-90f8-867dc9a2a6b0.png)

**Comment by Mukheem** [comment]
Result:
self = <test_pactmane2e.GetUserInfoContract testMethod=test_get_MS>
first = {'from': '2020-01-21T12:30:00+11:00', 'limit': 150, 'presenceChanges': [], 'skip': 0, ...}
second = {'from': '2020-01-21T12:30:00+11:00', 'limit': <pactman.mock.matchers.Like object at 0x0000023A79CAA400>, 'presenceChanges': [], 'skip': 0, ...}
msg = None

**Comment by r1chardj0n3s** [comment]
Hi, I can see a problem with your assertion `self.assertEqual(result.json(), expected)`: the value returned by the mock will be the concrete value supplied to the `Like()` - so in your example case, the mock will return `150`, so that assertion will fail. The correct assertion for your test case is `self.assertEqual(result.json(), { 'skip': 0, 'limit': 150 })`

Also, your `requests.get` appears to specify a real HTTP endpoint? It must use `pact.uri` base for URLs.

Also you're specifying a body and content-type for a GET request, which is not correct.

FWIW the complete working code I wrote to check your report is:
```import requests
import unittest

from pactman import Consumer, Provider, Like
pact = Consumer('Consumer').has_pact_with(Provider('Provider'))

class Test(unittest.TestCase):
    def test_get_MS(self):
        # Provider Start
        expected = { 'skip': 0, 'limit': Like(150)}
        pact.given(
            'User ML exists and is not an administrator'
        ).upon_receiving(
            'a request for ML'
        ).with_request(
            method='GET',
            path='/api',
        ).will_respond_with(200, body=expected )
        # Provider End
        
        # Consumer Start
        with pact:
            result = requests.get(pact.uri + '/api')

        self.assertEqual(result.json(), { 'skip': 0, 'limit': 150 })
        # Consumer End.


if __name__ == '__main__':
    unittest.main()
```

**Comment by Mukheem** [comment]
Hello @r1chardj0n3s ,

Thanks for your completed code. 

Here in the above code you are hitting the PACT server and getting the expected response then validating with a static actual value. This is contradicting with the example provided in your Documentation.
Kindly, refer the attached image.
![image](https://user-images.githubusercontent.com/19348206/74507469-81008a00-4f22-11ea-8314-c3669f025a5d.png)

I want to implement PACT as it is shown in the Image. Can you please help me with the LIKE matcher in this regard.

**Comment by LionelB5** [comment]
Hi @Mukheem,

Throughout the lifetime of the test case Richard provided in his response, there are no *actual* requests being made over a network.

I will try to explain with an example. Consider the following services:
- `users-service` (In this context, the **provider**)
- `stores-service` (In this context, the **consumer**)

I've written some code in the `users-service` that needs to make a call over the network to the `stores-service`.
Now I want to test this code, I could very easily use a library like `monkeypatch` to patch out the actual network call and mock what it would return. This does the job, but when I put my code into production, how will I know that the `users-service` will actually respond with a response body equivalent to what I have mocked the network call in the `stores-service` to return, or that it even supports the `/users/` endpoint?

Pact is a solution to the above.

When we declare a pact we are simply describing what the desired response should look like when the *consumer* makes a call to the *provider*, given a particular HTTP method, url and request body.

When the test case provided by Richard is executed, no actual HTTP calls will be made over the network, and the response from `get_user()` will always match what the `body` argument was set to when calling the `will_respond_with` method. Once the test is successfully executed, a JSON file will be created representing the interaction described by the pact used in the test case. This file should then be uploaded to a server running the [pact-broker](https://github.com/pact-foundation/pact_broker). It is now up to the provider to download the interactions registered against it from the pact broker, replay these interactions against itself, and publish the response to the pact-broker.

Say we've made some changes to the `stores-service`. We run our pact provider verification as part of our local tests, and see that some interactions don't match the expectations from our consumers, we now know we've made some breaking changes in our patch to the `stores-service`.

I will try and clarify the purpose of matchers such as `Like`. 

In the example you refer to in your image we are declaring an interaction which specifies that the response should be:
```
{
      'username': 'UserA',
      'id': 123,
      'groups': ['Editors']
}
```
When this pact is verified by the provider, the interaction will only be verified successfully if the provider produces a response containing at least the three keys, and their *exact* values as specified by the consumer.

For example, if the provider returned the following response when verifying the interaction, the interaction would not successfully verify:
```
{
      'username': 'UserA',
      'id': 123456,
      'groups': ['Editors']
}
```

The consumer might not care what the value of `id` actually is, it might just want to enforce that the `id` key is returned in the response, and that it has a value of type `int`. By using the `Like` matcher we are creating a pact that is not as strict. For example, if the consumer declared they were expecting a response like this:
```
{
      'username': 'UserA',
      'id': Like(123),
      'groups': ['Editors']
}
```
The pact would be verified successfully (by the provider), even if the provider returned a response that looks like this:
```
{
      'username': 'UserA',
      'id': 123456,
      'groups': ['Editors']
}
```
When the consumer runs its unit tests, if using the `Like` matcher like I have described above, the value returned by `get_user()` will always be:
```
{
      'username': 'UserA',
      'id': 123,
      'groups': ['Editors']
}
```

As Richard states, it doesn't really make sense to assert the equality of: 
```
{
      'username': 'UserA',
      'id': 123,
      'groups': ['Editors']
}
```
and
```
{
      'username': 'UserA',
      'id': Like(123),
      'groups': ['Editors']
}
```
as `Like(123)` (a `Like` object) is not actually equivalent to `123`. 

It's purpose is to generate a pact contract that is less strict.

I hope this helps!
